src/anyp/PortCfg.h

```diff
@@ -72,16 +80,72 @@ class PortCfg : public CodeContext
 
 private:
     explicit PortCfg(const PortCfg &other); // for ipV4clone() needs only!
+    char *name_; ///< internal name for the port
+    char *spec_; ///< the port specification (port or addr:port)
+};
+
+/// Iterates over a PortCfg list and sets the corresponding CodeContext before each iteration.
+class PortIterator
+{
+public:
+    // some of the standard iterator traits
+    using iterator_category = std::forward_iterator_tag;
+    using value_type = PortCfgPointer;
+    using pointer = value_type *;
+    using reference = value_type &;
+
+    /// \param first the PortCfg this iterator points to
+    PortIterator(const PortCfgPointer &first): position_(first) { setContext(); }
+
+    /// Special constructor for end() iterator.
+    /// This iterator does not point to an object and should never be dereferenced.
+    PortIterator(): position_(nullptr) {}
+
+    reference operator *() { return position_; }
+    pointer operator ->() { return &position_; }
+
+    PortIterator& operator++() { position_ = position_->next; setContext(); return *this; }
```

Comment:
I think an increment operator is the wrong place to reset context:

* We are missing setting context for the first port/iteration.
* We are forced to treat the last iteration step specially in setContext()

I suggest resetting context when the iterator is _dereferenced_ to obtain the loop value for the current iteration.

```C++
    reference operator *() { Assure(position_); CodeContext::Reset(position_); return *position_; }
```

Reply:
> However, in a common loop the code preceiding the initial dereferencing will still belong to the old context

This is one of those cases where my [precondition](https://github.com/measurement-factory/squid/pull/152#pullrequestreview-1031719849) plays a role: "In some of these change requests, I assume that we are targeting range-based for loops and, if there is a design conflict/choice, we optimize for those loops rather than other use cases."

When/if other use cases arise, we will either make PortCfgRange more complex to address them as well OR add new APIs that specialize in those other cases. For now, I would keep it as simple/straightforward as possible, optimizing for-range loop cases.